### PR TITLE
Fix bug in PDL encoding logic for imports when there are conflicting types in use.

### PR DESCRIPTION
**Pull request by karthikbalasub**
Unit tests added for coverage.

### CODE REVIEW

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 1]
```diff
@@ -719,6 +719,7 @@ private void writeDocAndProperties(RecordDataSchema.Field field) throws IOExcept
    * (1) The type is outside the root namespace of the document.
    * (2) The type is declared outside the document (i.e. not inlined in this document).
    * (3) The type's name does not conflict with name of an Inlined type.
+   * (4) The type's name does not conflict with another type that can use simple reference.
```
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: I'd say `Importing the type wouldn't force some intra-namespace reference to become fully-qualified.`

File: generator-test/src/test/pegasus/com/linkedin/pegasus/generator/test/idl/imports/ReferenceTypeConflict.pdl [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      /**
+       * This is a simple reference to the type matching the overriden namespace.
+       */
+      simpleReference: Fruits
+
+      /**
+       * Uses a fully qualified reference even if the type matches root namespace as this is within the overridden
+       * namespace.
+       */
+      fqnReference: com.linkedin.pegasus.generator.test.idl.imports.Fruits
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply generator-test/src/test/pegasus/com/linkedin/pegasus/generator/test/idl/imports/ReferenceTypeConflict.pdl]: good catch. Fixed

File: generator-test/src/test/pegasus/com/linkedin/pegasus/generator/test/idl/imports/ReferenceTypeConflict.pdl [hunk 3]
```diff
@@ -0,0 +1,40 @@
+namespace com.linkedin.pegasus.generator.test.idl.imports
+
+/**
+ * This record references two different Fruit enums, one in the same namespace, one from another namespace.
+ * The external namespace type should not be imported as it would conflict with the simple reference of the type from
+ * this record's namespace.
+ */
+record ReferenceTypeConflict {
+  /**
+   * This should be a simple reference as its namespace matches the root namespace.
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply generator-test/src/test/pegasus/com/linkedin/pegasus/generator/test/idl/imports/ReferenceTypeConflict.pdl]: fixed

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 4]
```diff
@@ -733,13 +734,17 @@ private void writeDocAndProperties(RecordDataSchema.Field field) throws IOExcept
   private Map<String, Name> computeImports(DataSchema schema, String rootNamespace)
   {
     Set<Name> encounteredTypes = new HashSet<>();
-    Set<String> inlinedTypeNames = new HashSet<>();
-    gatherTypes(schema, true, encounteredTypes, inlinedTypeNames);
+    // Collects the set of simple names of types that can use simple reference because
+    // 1. They are defined inline or
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: My comment was describing the same, made it clear now..

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 5]
```diff
@@ -821,10 +836,17 @@ else if (schema instanceof ArrayDataSchema)
...
   private void writeReference(NamedDataSchema schema) throws IOException
   {
-    if (schema.getNamespace().equals(_namespace) ||
-        (_importsByLocalName.containsKey(schema.getName()) &&
-        _importsByLocalName.get(schema.getName()).getNamespace().equals(schema.getNamespace())))
+    // Imports take precedence over current namespace
+    if (_importsByLocalName.containsKey(schema.getName()) &&
+        _importsByLocalName.get(schema.getName()).getNamespace().equals(schema.getNamespace()))
     {
+      // Write only identifier if there is an import matching the schema.
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: fixed

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 6]
```diff
@@ -821,10 +836,17 @@ else if (schema instanceof ArrayDataSchema)
...
+      _builder.writeIdentifier(schema.getName());
+    }
+    else if (_namespace.equals(schema.getNamespace()) && !_importsByLocalName.containsKey(schema.getName()))
+    {
+      // Write only identifier for schemas in the current namespace only if there are no imports with the same simple
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: fixed. 
corresponding imports -> conflicting imports

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 7]
```diff
@@ -821,10 +836,17 @@ else if (schema instanceof ArrayDataSchema)
    */
   private void writeReference(NamedDataSchema schema) throws IOException
   {
-    if (schema.getNamespace().equals(_namespace) ||
-        (_importsByLocalName.containsKey(schema.getName()) &&
-        _importsByLocalName.get(schema.getName()).getNamespace().equals(schema.getNamespace())))
+    // Imports take precedence over current namespace
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: fixed

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java (hunk 1 above)
Comment (karthikbalasub) [reply data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: fixed

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 8]
```diff
@@ -749,25 +754,29 @@ private void writeDocAndProperties(RecordDataSchema.Field field) throws IOExcept
...
   {
     // If named type, add to the set of encountered types
     if (schema instanceof NamedDataSchema)
     {
       NamedDataSchema namedSchema = (NamedDataSchema) schema;
       encounteredTypes.add(new Name(namedSchema.getFullName()));
-      // If declared inline, add to the set of inlined types
-      if (isDeclaredInline)
+      // If declared inline or of the namespace matches the current namespace, add to the set of types with simple
+      // reference.
```
[the remark this answers is not in the document]
Comment (karthikbalasub) [reply data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: fixed

File: data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java [hunk 9]
```diff
@@ -733,13 +735,18 @@ private void writeDocAndProperties(RecordDataSchema.Field field) throws IOExcept
   private Map<String, Name> computeImports(DataSchema schema, String rootNamespace)
   {
     Set<Name> encounteredTypes = new HashSet<>();
-    Set<String> inlinedTypeNames = new HashSet<>();
-    gatherTypes(schema, true, encounteredTypes, inlinedTypeNames);
+    // Collects the set of simple names of types that can cause conflicts with imports because
+    // 1. They are defined inline or
+    // 2. They are in the same namespace as their surrounding context (including namespace overrides) and are
+    //    preferred use simple reference
+    Set<String> typeNamesWithSimpleReference = new HashSet<>();
```
Comment (evanw555) [inline data/src/main/java/com/linkedin/data/schema/SchemaToPdlEncoder.java]: You should also name this to `nonImportableTypeNames` to be consistent.

**Pull request merged**